### PR TITLE
feature/ticket + some rubocop

### PR DESCRIPTION
**Pull request by Minpeople**
I created tickets model
I deleted reservation model
I made a unique relation between screenings and seats 
And added some rubocop ( I will ceeate new PR with rubocop.yml file and 'repair' more code style in next PR.

### CODE REVIEW

File: app/models/ticket.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class Ticket < ApplicationRecord
+  belongs_to :user
+  belongs_to :screening
+  STATUS = %w[pending reserved cancelled paid].freeze
+  TYPES = %w[Adult child student senior].freeze
```
Comment (kodowicz) [inline app/models/ticket.rb:7]: I would change `Adult` to downcase

Comment (kodowicz) [inline app/models/ticket.rb:7]: I would rewrite `Adult` as downcase

File: db/schema.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
```
Comment (kodowicz) [inline db/schema.rb:1]: `db/schema.rb` is autogenerated file and we shouldn't change that :)

File: Gemfile [hunk 3]
```diff
@@ -8,7 +8,7 @@ ruby '3.0.2'
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails', branch: 'main'
 gem 'rails', '~> 6.1.4'
 # Use sqlite3 as the database for Active Record
-#gem 'sqlite3', '~> 1.4'
+# gem 'sqlite3', '~> 1.4'
 group :development do
   gem 'sqlite3'
```
Comment (kodowicz) [inline Gemfile:13]: Why have you changed database? 🤔

Comment (Minpeople) [reply Gemfile:13]: @kodowicz
Since start I had sqlite 3, but for heroku I will need postgresql as production db, so I decided to change this

Comment (wmaciejak) [reply Gemfile:13]: Do I well understood that your app will work in production(heroku) env with PG, but in dev env, you will work with SQLite?

Never do that, please! You always should replicate your enviroment and it should be always consistent. Even if you have to spent few hours to do that, please migrate whole app to PostgreSQL also for local environment.

**Comment by wmaciejak** [comment]
> I created tickets model
> I deleted reservation model
> I made a unique relation between screenings and seats
> And added some rubocop ( I will ceeate new PR with rubocop.yml file and 'repair' more code style in next PR.

Remember that PR's are important from the perspective of understanding taken decisions. Could you describe why you deleted reservations at all? It's important cause I'm able to give you advice regarding architectural decisions when I understand your reasoning. In another way, I can only check Ruby syntax, nothing more.

**Comment by Minpeople** [comment]
> > I created tickets model
> > I deleted reservation model
> > I made a unique relation between screenings and seats
> > And added some rubocop ( I will ceeate new PR with rubocop.yml file and 'repair' more code style in next PR.
> 
> Remember that PR's are important from the perspective of understanding taken decisions. Could you describe why you deleted reservations at all? It's important cause I'm able to give you advice regarding architectural decisions when I understand your reasoning. In another way, I can only check Ruby syntax, nothing more.

@wmaciejak 
I drew a table of reservations and tickets (on a piece of paper) -their associations and attributes. I realized that they would duplicate, so I decided to standardize it in the form of the Ticket model. I just need TicketDesk (I thought about ActiveAdmin, but probably this wouldn't be the best choice) and Vouchers model.

**Comment by wmaciejak** [comment]
Okay got it, thank you @Minpeople. In this case, could you tell me how the process of buying 4 tickets will look like? Imagine that you going to cinema web and you choosing 4 tickets, then you will receive those tickets on your email. How this process from a technical perspective will look like?

**Comment by Minpeople** [comment]
> Okay got it, thank you @Minpeople. In this case, could you tell me how the process of buying 4 tickets will look like? Imagine that you going to cinema web and you choosing 4 tickets, then you will receive those tickets on your email. How this process from a technical perspective will look like?



> Okay got it, thank you @Minpeople. In this case, could you tell me how the process of buying 4 tickets will look like? Imagine that you going to cinema web and you choosing 4 tickets, then you will receive those tickets on your email. How this process from a technical perspective will look like?


Ok, so cinema employee log in using ticket desk. He choose screening, movie (which client wants) and he puts parametres of tickets - in this case 4. When it's done, app automatically sends email to client with his tickets. 
I understand I need a method to store more than one ticket before buying.

> Okay got it, thank you @Minpeople. In this case, could you tell me how the process of buying 4 tickets will look like? Imagine that you going to cinema web and you choosing 4 tickets, then you will receive those tickets on your email. How this process from a technical perspective will look like?

Ok, so cinema employee log in using ticket desk. He choose screening, movie (which client wants) and he puts parametres of tickets - in this case 4. When it's done, app automatically sends email to client with his tickets. 
I understand I need a method to store more than one ticket before buying or extend 'create' method. Probably I will create empty array and I will be adding there every 4 (from your example) tickets I generated

**Comment by wmaciejak** [comment]
>I understand I need a method to store more than one ticket before buying or extend 'create' method. Probably I will create empty array and I will be adding there every 4 (from your example) tickets I generate

I think that you should give the ability to just buy somehow more than one ticket (in one request).

I'm fine with the rest of your business workflow - looks correct. One thing that might be not possible without the `reservation` model will be canceling a whole bunch of tickets created in the scope of one "reservation". For instance, you have 4 tickets bought in one reservation and you will realize that you cannot participate and want to cancel the whole reservation - then it won't be possible probably. I'm fine with that if you don't see any blockers for you.

**Comment by Minpeople** [comment]
> > I understand I need a method to store more than one ticket before buying or extend 'create' method. Probably I will create empty array and I will be adding there every 4 (from your example) tickets I generate
> 
> I think that you should give the ability to just buy somehow more than one ticket (in one request).
> 
> I'm fine with the rest of your business workflow - looks correct. One thing that might be not possible without the `reservation` model will be canceling a whole bunch of tickets created in the scope of one "reservation". For instance, you have 4 tickets bought in one reservation and you will realize that you cannot participate and want to cancel the whole reservation - then it won't be possible probably. I'm fine with that if you don't see any blockers for you.

Ok, I tried to implement it in `create` method.

File: Gemfile (hunk 3 above)
Comment (Minpeople) [reply Gemfile]: I've migrated whole app to PostgreSQL :)

File: app/controllers/tickets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    reservation = [] 
+    
+    @ticket = Ticket.new(ticket_params).each do 
+      if @ticket.save
+        reservation << @ticket
+      else
+        render json: @tickets.errors, status: :unprocessable_entity
+      end
+    end
+    render json: reservation, status: :created
```
Comment (wmaciejak) [inline app/controllers/tickets_controller.rb:19]: How does this logic allow users to buy more than one ticket? Could you give examples of parameters that have to be sent here? I mean - how `ticket_params` should looks like?

Comment (Minpeople) [reply app/controllers/tickets_controller.rb:19]: @wmaciejak  Tickets have an array "reservation" and after creating each of tickets, they are in this array. So in ticket_params I have to add price. What is more: type of ticket, movie id, screening id and seats id

Comment (wmaciejak) [reply app/controllers/tickets_controller.rb:19]: Could add only this additional `seats_id`? Generally looks, fine, only wanted to confirm your way of thinking :)

**Review by wmaciejak — APPROVED** [review]

**Pull request merged**